paddle/fluid/pybind/imperative.cc

```diff
@@ -640,55 +640,109 @@ void BindImperative(py::module *m_ptr) {
+                   platform::errors::InvalidArgument(
+                       "Tensor %s has not been initialized!", self.Name()));
+
+               auto *detach_tensor =
+                   detach_var->MutableVar()->GetMutable<framework::LoDTensor>();
+               detach_tensor->ShareDataWith(origin_tensor);
+             } else {
+               const auto &origin_selected_rows =
+                   self.Var().Get<framework::SelectedRows>();
+               PADDLE_ENFORCE_EQ(
+                   origin_selected_rows.value().IsInitialized(), true,
+                   platform::errors::InvalidArgument(
+                       "Tensor %s has not been initialized!", self.Name()));
+
+               auto *detach_selected_rows =
+                   detach_var->MutableVar()
+                       ->GetMutable<framework::SelectedRows>();
+               detach_selected_rows->set_height(origin_selected_rows.height());
+               detach_selected_rows->set_rows(origin_selected_rows.rows());
+               detach_selected_rows->mutable_value()->ShareDataWith(
+                   origin_selected_rows.value());
+             }
+             VLOG(3) << "The detached Tensor(" << detach_var->Name()
+                     << ") share data with " << self.Name();
+             return detach_var;
            },
-           py::return_value_policy::copy, R"DOC(
+           py::return_value_policy::take_ownership, R"DOC(
 
         Returns a new Tensor, detached from the current graph.
+        It will share data with origin Tensor and always not have a Tensor copy and 
+        Tn addition, the detached Tensor does not provide gradient propagation.
 
         Returns: The detached Tensor.
 
         Examples:
             .. code-block:: python
 
                 import paddle
-                linear = Linear(32, 64)
-                data = paddle.uniform(shape=[30, 10, 32], -1, 1)
-                x = linear(data)
-                y = x.detach()
+
+                x = paddle.to_tensor(1.0, stop_gradient=False)
+                detach_x = x.detach()
+                detach_x[:] = 10.0
+                print(x.numpy())      # [10.0], detach_x share data with x
+
+                y = x**2
+                y.backward()
+                print(x.grad)         # [20.0]
+                print(detach_x.grad)  # None, 'stop_gradient=True' by default
+
+                detach_x.stop_gradient = False # Set stop_gradient to be False, supported auto-grad
+                z = detach_x**3
+                z.backward()
+
+                print(x.grad)         # [20.0], detach_x is detached from x's graph, not affect each other
+                print(detach_x.grad)  # [300.0], detach_x has its own graph
+
+                # Due to sharing of data with origin Tensor， There are some unsafe operations:
+                y = 2 * x
+                detach_x[:] = 5.0
+                y.backward() 
+                # It will raise Error:
```

Comment:
Error message may be need to modified by you. @liym27

Reply:
OK